### PR TITLE
More accurate handling of unicode/numpad input in gtk3 backends.

### PR DESCRIPTION
**Pull request by anntzer**
See changelog.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by QuLogic** [comment]
Interesting; some time back I tried to replace those magic numbers with GDK keysyms, but found that the numpad appeared to work backwards of numlock. I wonder if this will work correctly here.

**Comment by anntzer** [comment]
It works correctly locally, but CI seems unhappy...

File: doc/api/api_changes_3.4/behaviour.rst [hunk 1]
```diff
@@ -1,2 +1,15 @@
...
+~~~~~~~~~~~~~~~~~~~~
+
+The handling of non-ascii keypresses (as reported in the KeyEvent passed to
+``key_press_event``-handlers) in the GTK backends now correctly reports unicode
+characters (e.g., €), and respects NumLock on the numpad.
+
+The following key names have changed; the new names are consistent with those
+reported by the Qt backends:
+- The "Break/Pause" key (keysym 0xff13) is now reported as "pause" instead of
+  "break" (this is also consistent with the X key name).
```
Comment (timhoffm) [inline doc/api/api_changes_3.4/behaviour.rst:14]: For some reason sphinx is unhappy here:

> /home/circleci/project/doc/api/api_changes_3.4/behaviour.rst:14: WARNING: Unexpected indentation.
/home/circleci/project/doc/api/api_changes_3.4/behaviour.rst:15: WARNING: Block quote ends without a blank line; unexpected 

Maybe one cannot continue the list item with double quotes? Try moving "of" to this line.

Comment (anntzer) [reply doc/api/api_changes_3.4/behaviour.rst:15]: I think I just need a blank line before the list starts.

**Comment by QuLogic** [comment]
CI doesn't like your test, it seems.

**Comment by anntzer** [comment]
I know :/

**Comment by QuLogic** [comment]
This fixes input via numpad for me.

**Comment by anntzer** [comment]
I decided to just skip the gtk test on travis for now...

File: lib/matplotlib/tests/test_backend_gtk3.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                 Gdk.ModifierType.CONTROL_MASK
+                 | Gdk.ModifierType.MOD1_MASK
+                 | Gdk.ModifierType.MOD4_MASK),
+                (0xfd16, 0),   # KEY_3270_Play.
+                (Gdk.KEY_BackSpace, 0),
+                (Gdk.KEY_BackSpace, Gdk.ModifierType.CONTROL_MASK),
+        ]:
+            # This is not actually really the right API: it depends on the
+            # actual keymap (e.g. on Azerty, shift+agrave -> 0).
+            Gtk.test_widget_send_key(fig.canvas, key, mod)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_gtk3.py:38]: Could this be why it's getting stuck? Maybe you can check `len(buf)` and assert the result after shutting down?

File: doc/api/next_api_changes/behavior/17791-AL.rst [hunk 3]
```diff
@@ -0,0 +1,13 @@
+GTK key name changes
+~~~~~~~~~~~~~~~~~~~~
+
+The handling of non-ascii keypresses (as reported in the KeyEvent passed to
+``key_press_event``-handlers) in the GTK backends now correctly reports unicode
```
[the remark this answers is not in the document]
Comment (anntzer) [reply doc/api/next_api_changes/behavior/17791-AL.rst]: yup

File: lib/matplotlib/tests/test_backend_gtk3.py (hunk 2 above)
Comment (anntzer) [reply lib/matplotlib/tests/test_backend_gtk3.py:35]: Actually turns out that even locally this is not passing anymore, i.e. calls to Gtk.test_widget_send_event (or Gdk.simulate_key_event) don't result in the key_press_event handler being called at all.  At some point it worked, so likely a Gtk version thing?  I just xfailed the whole thing for now...

Comment (QuLogic) [reply lib/matplotlib/tests/test_backend_gtk3.py:35]: According to https://mail.gnome.org/archives/commits-list/2017-January/msg01173.html, those were never fully implemented (I only see an x11 version in that commit).

**Review by QuLogic — APPROVED** [review]

**Comment by anntzer** [comment]
rebased

**Review by timhoffm — APPROVED** [review]

**Pull request merged**